### PR TITLE
[TOPI] Conv2d Added and Optimized for Intel HD Graphics

### PR DESCRIPTION
**Pull request by Laurawly**
Optimize convolutions specifically for Intel Graphics which compared with other opencl based backends, adds specific intel_sub_group operations. And we support consistent data packing (thanks to @yzhliu's implementation ) for end-to-end networks. Note that currently conv2d is most optimized for workloads for resnet. We compared with state-of-the-art deep learning libraries for intel GPUs. Following is some end-to-end benchmark for Intel® Processor Graphics Gen9 using opt_level 3.

| Networks | TVM  | clDNN (Intel's deep learning library)  |
| ---------  | ----- | ------------------------------------ |
| ResNet-18 | 74ms | 70ms |
| ResNet-50 | 262 ms | 208.4 ms|

### CODE REVIEW

**Comment by tqchen** [comment]
@yidawang Why not intel_gpu? seems intel_graphics is a bit too long for abbreviation.

**Comment by Laurawly** [comment]
Please feel free to review and comment @yzhliu @Huyuwei

**Comment by yidawang** [comment]
@tqchen I feel that "Intel GPU" is a coined name, Intel normally calls it "Intel Graphics". And, Intel plans to launch their first version of discrete GPU, which will be the real "Intel GPU", in 2020. So I think it is helpful to reduce the confusion from the beginning.

**Comment by kevinthesun** [comment]
It would be nice if later we can parameterize schedules, like conv2d for x86.

**Comment by liangfu** [comment]
How about 'Iris'? Intel brand its high end graphics under the name "iris plus".

**Comment by merrymercy** [comment]
We will rewrite all performance bottleneck operators (e.g. conv2d)  in autotuning fashion after merging the auto tuner. For now it is good to follow @kevinthesun 's comment .

File: topi/python/topi/intel_graphics/conv2d.py [hunk 1]
```diff
@@ -0,0 +1,557 @@
...
+    thread_x = tvm.thread_axis((0, x_factor), "threadIdx.x")
+    s[tensor].bind(zo, tvm.thread_axis("blockIdx.z"))
+    s[tensor].bind(zi, thread_z)
+    s[tensor].bind(yo, tvm.thread_axis("blockIdx.y"))
+    s[tensor].bind(yi, thread_y)
+    s[tensor].bind(xo, tvm.thread_axis("blockIdx.x"))
+    s[tensor].bind(xi, thread_x)
+    return xi, thread_z, thread_y, thread_x
+
+@conv2d_alter_layout.register(["intel_gpu"])
```
Comment (tqchen) [inline topi/python/topi/intel_graphics/conv2d.py]: consider making the keys consistent as per @yidawang 's suggestion

File: topi/python/topi/intel_graphics/conv2d.py [hunk 2]
```diff
@@ -0,0 +1,557 @@
...
+        HSTR, WSTR = stride
+    else:
+        HSTR, WSTR = stride, stride
+
+    return _decl_cl_spatialpack(data, kernel, stride, padding, layout, out_dtype)
+
+@generic.schedule_conv2d_nchw.register(["intel_gpu"])
+def schedule_conv2d_nchw(outs):
+    """Schedule for conv2d_nchw for Intel GPU
+    Parameters
```
Comment (tqchen) [inline topi/python/topi/intel_graphics/conv2d.py]: there needs to be a new line between Parameters and usage.

See https://docs.tvm.ai/contribute/document.html#document-python

**Comment by tqchen** [comment]
@Laurawly please act on the comments and let us bring the changes in

**Comment by Laurawly** [comment]
@tqchen So we should change all intel_gpu to intel_graphics including the target naming in tvm right?

File: include/tvm/build_module.h [hunk 3]
```diff
@@ -157,8 +157,8 @@ EXPORT Target rasp(const std::vector<std::string>& options =
 EXPORT Target mali(const std::vector<std::string>& options =
                    std::vector<std::string>());
 
-/*! \return A target for Intel GPU */
-EXPORT Target intel_gpu(const std::vector<std::string>& options =
+/*! \return A target for Intel Graphics */
+EXPORT Target intel_graphics(const std::vector<std::string>& options =
                    std::vector<std::string>());
```
Comment (tqchen) [inline include/tvm/build_module.h]: alignment

**Comment by tqchen** [comment]
@yidawang please review again and approve or request more changes

**Comment by yidawang** [comment]
@tqchen I agree with @kevinthesun and @merrymercy  about parameterizing the schedules like https://github.com/dmlc/tvm/blob/master/topi/python/topi/x86/conv2d.py#L15 But I think we can do it in a separated PR later.

Other than that, LGTM

**Comment by tqchen** [comment]
@yidawang please explicit approve, see https://docs.tvm.ai/contribute/code_review.html#approve-and-request-changes-explicitly

**Pull request merged**